File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/logic/generic_commander_logic.cpp
#include "crtp_cpp/logic/generic_commander_logic.hpp"

GenericCommanderLogic::GenericCommanderLogic(CrtpLink* crtp_link)
    : Logic(crtp_link),
      packer(GenericCommanderPacker()) {}

void GenericCommanderLogic::send_notify_setpoints_stop(uint32_t remain_valid_milliseconds) {
    CrtpRequest request;
    request.packet = packer.send_notify_sendpoints_stop(remain_valid_milliseconds);
    link->send_packet_no_response(request);
}

void GenericCommanderLogic::send_stop_setpoint() {
    CrtpRequest request;
    request.packet = packer.send_stop_setpoint();
    link->send_packet_no_response(request);
}

void GenericCommanderLogic::send_velocity_world_setpoint(float vx, float vy, float vz, float yawrate) {
    CrtpRequest request;
    request.packet = packer.send_velocity_world_setpoint(vx, vy, vz, yawrate);
    link->send_packet_no_response(request);
}

void GenericCommanderLogic::send_zdistance_setpoint(float roll, float pitch, float yawrate, float zdistance) {
    CrtpRequest request;
    request.packet = packer.send_zdistance_setpoint(roll, pitch, yawrate, zdistance);
    link->send_packet_no_response(request);
}

void GenericCommanderLogic::send_hover_setpoint(float vx, float vy, float yawrate, float zdistance) {
    CrtpRequest request;
    request.packet = packer.send_hover_setpoint(vx, vy, yawrate, zdistance);
    link->send_packet_no_response(request);
}

void GenericCommanderLogic::send_full_state_setpoint(
    const std::vector<float>& pos, const std::vector<float>& vel, const std::vector<float>& acc,
    const std::vector<float>& orientation, float rollrate, float pitchrate, float yawrate) {

    auto vector_to_mm_16bit = [](const std::vector<float>& vec) {
        return std::make_tuple(static_cast<int16_t>(vec[0] * 1000), static_cast<int16_t>(vec[1] * 1000), static_cast<int16_t>(vec[2] * 1000));
    };

    int16_t x, y, z;
    std::tie(x, y, z) = vector_to_mm_16bit(pos);
    int16_t vx, vy, vz;
    std::tie(vx, vy, vz) = vector_to_mm_16bit(vel);
    int16_t ax, ay, az;
    std::tie(ax, ay, az) = vector_to_mm_16bit(acc);
    int16_t rr, pr, yr;
    std::tie(rr, pr, yr) = vector_to_mm_16bit({rollrate, pitchrate, yawrate});
    
    float q[4] = { orientation[0], orientation[1], orientation[2], orientation[3] };
    uint32_t orient_comp = compress_quaternion(q); 

    CrtpRequest request;
    request.packet = packer.send_full_state_setpoint(x, y, z, vx, vy, vz, ax, ay, az, orient_comp, rr, pr, yr);
    link->send_packet_no_response(request);
}

void GenericCommanderLogic::send_position_setpoint(float x, float y, float z, float yaw) {
    CrtpRequest request;
    request.packet = packer.send_position_setpoint(x, y, z, yaw);
    link->send_packet_no_response(request);
}

uint32_t GenericCommanderLogic::compress_quaternion(float const q[4]) {
    // we send the values of the quaternion's smallest 3 elements.
    unsigned i_largest = 0;
    for (unsigned i = 1; i < 4; ++i) {
        if (fabsf(q[i]) > fabsf(q[i_largest])) {
        i_largest = i;
        }
    }

    // since -q represents the same rotation as q,
    // transform the quaternion so the largest element is positive.
    // this avoids having to send its sign bit.
    unsigned negate = q[i_largest] < 0;

    // 1/sqrt(2) is the largest possible value
    // of the second-largest element in a unit quaternion.

    // do compression using sign bit and 9-bit precision per element.
    uint32_t comp = i_largest;
    for (unsigned i = 0; i < 4; ++i) {
        if (i != i_largest) {
        unsigned negbit = (q[i] < 0) ^ negate;
        unsigned mag = ((1 << 9) - 1) * (fabsf(q[i]) / (float)M_SQRT1_2) + 0.5f;
        comp = (comp << 10) | (negbit << 9) | mag;
        }
    }

    return comp;
}

File: src/dependencies/crazyflie_hardware/src/crtp/crtp/logic/generic_commander_logic.py
from .logic import Logic
from crtp.crtp_link import CrtpLink
from crtp.packers.generic_commander_packer import GenericCommanderPacker
from crtp.packers.crtp_packer import CrtpPacker

from typing import Callable

from crtp.utils.encoding import compress_quaternion


class GenericCommanderLogic(Logic):
    """
    Used for sending control setpoints to the Crazyflie
    """

    def __init__(
        self, crtp_packer_factory: Callable[[int], CrtpPacker], crtp_link: CrtpLink
    ):
        super().__init__(crtp_link)
        self.packer = GenericCommanderPacker(crtp_packer_factory)

    def send_notify_setpoints_stop(self, remain_valid_milliseconds=0):
        """
        Sends a packet so that the priority of the current setpoint to the lowest non-disabled value,
        so any new setpoint regardless of source will overwrite it.
        """
        packet = self.packer.send_notify_sendpoints_stop(remain_valid_milliseconds)
        self.link.send_packet_no_response(packet)

    def send_stop_setpoint(self):
        """
        Send STOP setpoing, stopping the motors and (potentially) falling.
        """
        packet = self.packer.send_stop_setpoint()
        self.link.send_packet_no_response(packet)

    def send_velocity_world_setpoint(self, vx, vy, vz, yawrate):
        """
        Send Velocity in the world frame of reference setpoint with yawrate commands

        vx, vy, vz are in m/s
        yawrate is in degrees/s
        """
        packet = self.packer.send_velocity_world_setpoint(vx, vy, vz, yawrate)
        self.link.send_packet_no_response(packet)

    def send_zdistance_setpoint(self, roll, pitch, yawrate, zdistance):
        """
        Control mode where the height is send as an absolute setpoint (intended
        to be the distance to the surface under the Crazflie), while giving roll,
        pitch and yaw rate commands

        roll, pitch are in degrees
        yawrate is in degrees/s
        zdistance is in meters
        """
        packet = self.packer.send_zdistance_setpoint(roll, pitch, yawrate, zdistance)
        self.link.send_packet_no_response(packet)

    def send_hover_setpoint(self, vx, vy, yawrate, zdistance):
        """
        Control mode where the height is send as an absolute setpoint (intended
        to be the distance to the surface under the Crazflie), while giving x, y velocity
        commands in body-fixed coordinates.

        vx,  vy are in m/s
        yawrate is in degrees/s
        zdistance is in meters
        """
        packet = self.packer.send_hover_setpoint(vx, vy, yawrate, zdistance)
        self.link.send_packet_no_response(packet)

    def send_full_state_setpoint(
        self, pos, vel, acc, orientation, rollrate, pitchrate, yawrate
    ):
        """
        Control mode where the position, velocity, acceleration, orientation and angular
        velocity are sent as absolute (world) values.

        position [x, y, z] are in m
        velocity [vx, vy, vz] are in m/s
        acceleration [ax, ay, az] are in m/s^2
        orientation [qx, qy, qz, qw] are the quaternion components of the orientation
        rollrate, pitchrate, yawrate are in degrees/s
        """

        def vector_to_mm_16bit(vec):
            return int(vec[0] * 1000), int(vec[1] * 1000), int(vec[2] * 1000)

        x, y, z = vector_to_mm_16bit(pos)
        vx, vy, vz = vector_to_mm_16bit(vel)
        ax, ay, az = vector_to_mm_16bit(acc)
        rr, pr, yr = vector_to_mm_16bit([rollrate, pitchrate, yawrate])
        orient_comp = compress_quaternion(orientation)
        packet = self.send_full_state_setpoint(
            x, y, z, vx, vy, vz, ax, ay, az, orient_comp, rr, pr, yr
        )
        self.link.send_packet_no_response(packet)

    def send_position_setpoint(self, x, y, z, yaw):
        """
        Control mode where the position is sent as absolute (world) x,y,z coordinate in
        meter and the yaw is the absolute orientation.

        x, y, z are in m
        yaw is in degrees
        """
        packet = self.packer.send_position_setpoint(x, y, z, yaw)
        self.link.send_packet_no_response(packet)
